clang/runtime/dpct-rt/include/sparse_utils.hpp.inc

```diff
@@ -78,6 +79,241 @@ private:
   oneapi::mkl::index_base _index_base = oneapi::mkl::index_base::zero;
 };
 // DPCT_LABEL_END
+
+// DPCT_LABEL_BEGIN|csrmv|dpct::sparse
+// DPCT_DEPENDENCY_BEGIN
+// SparseUtils|matrix_info
+// Util|DataType
+// LibCommonUtils|get_value
+// LibCommonUtils|get_memory
+// DPCT_DEPENDENCY_END
+// DPCT_CODE
+/// Computes a CSR format sparse matrix-dense vector product.
+/// y = alpha * op(A) * x + beta * y
+/// \param [in] queue The queue where the routine should be executed.
+/// \param [in] trans The operation applied to the matrix A.
+/// \param [in] num_rows Number of rows of the matrix A.
+/// \param [in] num_cols Number of columns of the matrix A.
+/// \param [in] alpha Scaling factor for the matrix A.
+/// \param [in] info Matrix info of the matrix A.
+/// \param [in] val An array containing the non-zero elements of the matrix A.
+/// \param [in] row_ptr An array of length \p num_rows + 1.
+/// \param [in] col_ind An array containing the column indices in index-based
+/// numbering.
+/// \param [in] x Data of the vector x.
+/// \param [in] beta Scaling factor for the vector x.
+/// \param [in, out] y Data of the vector y.
+template <typename T>
+void csrmv(sycl::queue &queue, oneapi::mkl::transpose trans, int num_rows,
+           int num_cols, const T *alpha,
+           const std::shared_ptr<matrix_info> info, const T *val,
+           const int *row_ptr, const int *col_ind, const T *x, const T *beta,
+           T *y) {
+  using Ty = typename dpct::DataType<T>::T2;
+  auto alpha_value =
+      detail::get_value(reinterpret_cast<const Ty *>(alpha), queue);
```

Comment:
are alpha and beta possibly asynchronously having values set?  or is it safe to always access them here in the synchronous host mode.

Reply:
the detail::get_value() will try to use q.memcpy.wait() to copy alpha/beta to host if it is a device pointer.